### PR TITLE
Adding a debug only config knob for testing loop unrolling limits

### PR DESCRIPTION
**Pull request by tannergooding**
This will allow easy testing of different loop unroll counts by simply specifying `COMPlus_JitUnrollLoopMaxIterationCount=#`

The AMD, Arm, and Intel architecture manuals all recommend unrolling small loops. However, the limit and size of these loops can differ based on CPU. All of them agree on 4-8 for "small" loops and so I expect this may be a good starting point for doing some benchmarks.

CC. @BruceForstall

### CODE REVIEW

File: src/coreclr/jit/optimizer.cpp [hunk 1]
```diff
@@ -4201,12 +4201,15 @@ PhaseStatus Compiler::optUnrollLoops()
...
             unrollLimitSz = INT_MAX;
         }
-        else if (!(loopFlags & LPFLG_SIMD_LIMIT))
+        else if ((loopFlags & LPFLG_SIMD_LIMIT) == 0)
         {
-            // Otherwise unroll only if limit is Vector_.Length
-            // (as a heuristic, not for correctness/structural reasons)
-            JITDUMP("Failed to unroll loop " FMT_LP ": constant limit isn't Vector<T>.Length (heuristic)\n", lnum);
-            continue;
+            if (loop.lpConstLimit() > opts.compJitUnrollLoopIterationLimit)
```
Comment (BruceForstall) [inline src/coreclr/jit/optimizer.cpp:4206]: I think you want to be comparing against `totalIter`.

Maybe, for better clarity, the code should be:
```
else if (totalIter <= 1)
{
...
}
else if (totalIter <= opts.compJitUnrollMaxIterationCount)
{
  // We can unroll this
}
else if ((loopFlags & LPFLG_SIMD_LIMIT) != 0)
{
  // We can unroll this
}
else
{
  JITDUMP("Failed to unroll loop " FMT_LP ": insufficiently simple loop (heuristic)\n", lnum);
  continue;
}
```

**Review by BruceForstall — CHANGES_REQUESTED** [review]

File: src/coreclr/jit/compiler.h [hunk 2]
```diff
@@ -9303,6 +9303,9 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 // likely complicated enough that loop alignment will not impact performance.
 #define DEFAULT_MAX_LOOPSIZE_FOR_ALIGN DEFAULT_ALIGN_LOOP_BOUNDARY * 3
 
+// By default only single iteration loops will be unrolled
+#define DEFAULT_UNROLL_LOOP_LIMIT 1
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply src/coreclr/jit/compiler.h:9307]: Updated

File: src/coreclr/jit/optimizer.cpp (hunk 1 above)
Comment (tannergooding) [reply src/coreclr/jit/optimizer.cpp:4206]: Updated.

File: src/coreclr/jit/jitconfigvalues.h [hunk 3]
```diff
@@ -76,6 +76,8 @@ CONFIG_INTEGER(JitHideAlignBehindJmp,
 CONFIG_INTEGER(JitOptimizeStructHiddenBuffer, W("JitOptimizeStructHiddenBuffer"), 1) // Track assignments to locals done
                                                                                      // through return buffers.
 
+CONFIG_INTEGER(JitUnrollLoopLimit, W("JitUnrollLoopLimit"), DEFAULT_UNROLL_LOOP_MAX_ITERATION_COUNT)
```
Comment (BruceForstall) [inline src/coreclr/jit/jitconfigvalues.h:79]:
```suggestion
CONFIG_INTEGER(JitUnrollLoopMaxIterationCount, W("JitUnrollLoopMaxIterationCount"), DEFAULT_UNROLL_LOOP_MAX_ITERATION_COUNT)
```

File: src/coreclr/jit/optimizer.cpp [hunk 4]
```diff
@@ -4201,12 +4201,18 @@ PhaseStatus Compiler::optUnrollLoops()
...
-            continue;
+          // We can unroll this
+        }
+        else if ((loopFlags & LPFLG_SIMD_LIMIT) != 0)
+        {
+          // We can unroll this
+        }
+        else
+        {
+          JITDUMP("Failed to unroll loop " FMT_LP ": insufficiently simple loop (heuristic)\n", lnum);
```
Comment (BruceForstall) [inline src/coreclr/jit/optimizer.cpp:4214]: I'm guessing the formatter will complain about insufficient indenting; make sure you run jit-format before pushing the PR.

**Review by BruceForstall — APPROVED** [review]

**Pull request merged**